numpy/core/src/multiarray/refcount.c.src

```diff
@@ -14,46 +14,113 @@
- * such as as `np.dtype([("field1", "O"), ("field2", "f,O", (3,2))])`
+
+/**begin repeat
+ * #name = INCREF, XDECREF, CLEAR, VISIT#
+ * #inner_function = _PyArray_Item_INCREF, _PyArray_Item_XDECREF,
+ *                   _PyArray_Item_CLEAR, _PyArray_Item_VISIT#
+ * #is_visit_func = 0, 0, 0, 1#
+ * #item_func = Py_XINCREF, Py_XDECREF,, Py_VISIT#
+ * #set_to_none = 0, 0, 1, 0#
+ * #silence_errors = 0, 0, 1, 1#
+ * #context = "", "", "<ndarray.tp_clear>", "<ndarray.tp_traverse>"#
  */
-NPY_NO_EXPORT void
-PyArray_Item_INCREF(char *data, PyArray_Descr *descr)
+
+/* Use macros to achieve the different behaviour and function signatures */
+#if !@set_to_none@
+   #define HANDLE_ITEM_AT(data) @item_func@(*(data))
+#else
+    #define HANDLE_ITEM_AT(data) Py_XSETREF(*(data), Py_None)
+#endif
+
+
+#define HANDLE_ITEM_AT_UNALIGNED(data)          \
+    do {                                        \
+        PyObject *_tmp;                         \
+        NPY_COPY_PYOBJECT_PTR(&_tmp, data);     \
+        HANDLE_ITEM_AT(&_tmp);                  \
+        if (@set_to_none@) {                    \
+            NPY_COPY_PYOBJECT_PTR(data, &_tmp); \
+        }                                       \
+    } while (0)
+
+
+#if @is_visit_func@
+    #define CALL_INNER_FUNCTION(data, descr) \
+        @inner_function@(data, descr, visit, arg)
+    // #define ERR_RETURN PyErr_WriteUnraisable(Py_None); return 0
+#else
+    #define CALL_INNER_FUNCTION(data, descr) @inner_function@(data, descr)
+    // #define ERR_RETURN return -1;
+#endif
+
+#if @silence_errors@
+    #define ERROR_RETURN                                            \
+        do {                                                        \
+            PyObject *_err_string = PyUString_FromString(@context@);\
+            if (_err_string) {                                      \
+                PyErr_WriteUnraisable(_err_string);                 \
+                Py_DECREF(_err_string);                             \
+            }                                                       \
+            else {                                                  \
+                PyErr_WriteUnraisable(Py_None);                     \
+            }                                                       \
+            /* Only Py_VISIT returns not 0 for the VISIT func */    \
+            return 0;                                               \
+        } while (0)
+#else
+    #define ERROR_RETURN return -1
+#endif
+
+/*
+ * Function visiting all elements of object type within a given datatype.
+ * This is for structured dtypes and the inner function may be called
+ * recursively to visit all sub structures.
+ * If this is the VISIT function, it needs special arguments for Python
+ */
+NPY_NO_EXPORT int 
+ #if @is_visit_func@
+    @inner_function@(char *data, PyArray_Descr *descr,
+                     visitproc visit, void *arg)
+#else
+    @inner_function@(char *data, PyArray_Descr *descr)
+#endif
```

Comment:
This is really hard to reason about. Maybe it would be better to expand some of these templates at the cost of some code repitition.

Reply:
changed slightly, not sure it is much better with inline `#defines` for the arguments though.